0000-modules/detailed-design/visibility-and-reexport.md

```diff
@@ -0,0 +1,174 @@
+`pub` instead of `export`, and if they mark something `export`, they *know* it
+is part of their public API.
+
+Additionally, it will become more meaningful to search a crate for all
+instances of `export` than it today is to search for all instances of `pub`.
+After this change, that search will tend to give you a good sense of the entire
+API this crate exposes, whereas today it includes many items which are marked
+`pub` but are not exposed.
+
+Finally, though redefining `pub` may seem like a very significant breakage,
+practical breakages will be limited by the [migration][migration] mechanism.
+Repurposing the `pub` syntax will also limit some of the "long tail" effects of
+breakages. For example, if outdated documentation or Q&A content uses the `pub`
+visibility in a code snippet, that code will continue to compile, whereas if a
+new keyword were used, that code would not work on a newer Rust checkpoint.
+
+The distinction between the two definitions of `pub` really only impacts
+library authors who are in the process of migrating across versions. In
+binaries, or even parts of libraries that aren't part of the external API, the
+change in the meaning of `pub` does not make code stop working.
+
+### New re-export mechanism
+
+We also introduce a new mechanism for re-export, having the syntax:
+
+```rust
+$visibility $path;
+```
+
+This is similar to the `pub use` system we have today, but the `use` keyword is
+omitted.
+
+Other than the syntactic difference, another big difference is that these paths
+are relative paths, the same paths used everywhere inside of modules (except
+for `use` statements).
+
+Most commonly, the visibility used will be the export visibility. When you
+compare a facaded module like `futures::future`, for example, it will go from
+this:
+
+```rust
+mod and_then;
+mod map;
+mod select;
+mod then;
+
+pub use self::and_then::AndThen;
+pub use self::map::Map;
+pub use self::select::Select;
+pub use self::then::Then;
+// etc
+```
+
+To this:
+
+```rust
+export and_then::AndThen;
+export map::Map;
+export select::Select;
+export then::Then;
+```
+
+#### Exporting modules
+
+Because the blockless mod statement is deprecated, we need a new mechanism for
+making modules a part of the public API. In this system, modules are made
+public by exporting them:
+
+```rust
+// Exports the `foo` module
+export foo;
+```
+
+In this way, the syntax for mounting a module which is another file at this
+location in the public API is the same as the syntax for mounting any other
+item from other files at this point in the public API.
+
+#### Private-in-public & re-exporting
+
+It is not permissible to use this mechanism to re-export something at a greater
+visibility than its explicitly declared visibility. That is, this would be
+invalid:
+
+```
+mod foo {
+    pub struct Foo;
+}
+
+// This is an error: you have exported something which is only marked `pub`
+export foo::Foo;
+```
```

Comment:
To clarify, this should work:

```rust
mod foo {
    export struct Foo;
}
export foo::Foo;
```
but removing `export foo::Foo;` would cause a hard error. Is that correct?